nmdc_runtime/site/ops.py

```diff
@@ -1043,18 +1046,200 @@ def site_code_mapping() -> dict:
         )
 
 
+def _add_related_ids_to_alldocs(
+    temp_collection, context, document_reference_ranged_slots_by_type
+):
+    """
+    Adds {`_inbound`,`_outbound`} fields to each document in the temporary alldocs collection.
+
+    The {`_inbound`,`_outbound`} fields each contain an array of subdocuments, each with fields `id` and `type`.
+    Each subdocument represents a link to any other document that either links to or is linked from
+    the document via document-reference-ranged slots.
+
+    Args:
+        temp_collection: The temporary MongoDB collection to process
+        context: The Dagster execution context for logging
+        document_reference_ranged_slots_by_type: Dictionary mapping document types to their reference-ranged slot names
+
+    Returns:
+        None (modifies the documents in place)
+    """
+
+    context.log.info(
+        "Building relationships and adding `_inbound` and `_outbound` fields..."
+    )
+
+    # document ID -> type (with "nmdc:" prefix preserved)
+    id_to_type_map: Dict[str, str] = {}
+
+    # set of (<referencing document ID>, <slot>, <referenced document ID>) 3-tuples.
+    relationship_triples: Set[Tuple[str, str, str]] = set()
+
+    # Collect relationship triples.
+    for doc in temp_collection.find():
+        doc_id = doc["id"]
+        # Store the full type with prefix intact
+        doc_type = doc["type"]
+        # For looking up reference slots, we still need the type without prefix
+        doc_type_no_prefix = doc_type[5:] if doc_type.startswith("nmdc:") else doc_type
+
+        # Record ID to type mapping - preserve the original type with prefix
+        id_to_type_map[doc_id] = doc_type
+
+        # Find all document references from this document
+        reference_slots = document_reference_ranged_slots_by_type.get(
+            doc_type_no_prefix, []
+        )
+        for slot in reference_slots:
+            if slot in doc:
+                # Handle both single-value and array references
+                refs = doc[slot] if isinstance(doc[slot], list) else [doc[slot]]
+                for ref_doc in temp_collection.find(
+                    {"id": {"$in": refs}}, ["id", "type"]
+                ):
+                    id_to_type_map[ref_doc["id"]] = ref_doc["type"]
+                for ref_id in refs:
+                    relationship_triples.add((doc_id, slot, ref_id))
+
+    context.log.info(
+        f"Found {len(id_to_type_map)} documents, with "
+        f"{len({d for (d, _, _) in relationship_triples})} containing references"
+    )
+
+    # The bifurcation of document-reference-ranged slots as "inbound" and "outbound" is essential
+    # in order to perform graph traversal and collect all entities "related" to a given entity without
+    # recursion "exploding".
+    #
+    # An "inbound" slot is one for which an entity in the domain prov:wasInfluencedBy an entity in the range.
```

Comment:
When I first read this comment, I thought "prov:wasInfluencedBy" was the name of the domain, and then I thought a verb was missing after that (I was having trouble parsing the sentence). I recommend rephrasing it to be more similar to the "inverse" comment on lines 1124-1125. For example:

```suggestion
    # An "inbound" slot is one for which an entity in the domain is "influenced by"
    # (i.e., prov:wasInfluencedBy) an entity in the range.
```